### PR TITLE
[23.0] Workflow connections performance

### PR DESCRIPTION
**Pull request by ElectronicBlueberry**
Replaces the SVG based connections renderer with a canvas based one, significantly improving performance on workflows with many connections.

The connections hover style was removed as a consequence of this change, as calculating the hover over a curve is non-trivial.

For testing this feature, I initially tested it with single stroke lines, but became so attached to this look, that I included it in this PR. An upside is that single stroke curves require a bit less rendering resources. Let me know what you think about this style of connector.

This is what the connections currently look like:

image

image

image

image

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by ElectronicBlueberry** [comment]
Selenium failures are related. Working on a fix

File: client/src/components/Workflow/Editor/WorkflowEdges.vue [hunk 1]
```diff
@@ -1,50 +1,86 @@
...
+    if (!ctx) {
+        return;
+    }
+
+    ctx.canvas.width = ctx.canvas.offsetWidth;
+    ctx.canvas.height = ctx.canvas.offsetHeight;
+
+    ctx.resetTransform();
+    ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height);
+    ctx.translate(props.transform.x, props.transform.y);
```
Comment (mvdbeek) [inline client/src/components/Workflow/Editor/WorkflowEdges.vue:64]: That is awesome, it'll be done just once for the canvas, right ?

Comment (ElectronicBlueberry) [reply client/src/components/Workflow/Editor/WorkflowEdges.vue:64]: On every animation frame, once for the entire canvas.

**Comment by ElectronicBlueberry** [comment]
This is at a state I'm happy with.
I cleaned the commit history a bit. The failing tests appear to be unrelated.

File: lib/galaxy_test/selenium/test_workflow_editor.py [hunk 2]
```diff
@@ -1052,17 +1052,22 @@ def workflow_editor_connect(self, source, sink, screenshot_partial=None):
...
+    def assert_connected(self, _source, _sink):
+        # NOOP
+        # Used to assert that an svg connection element exists.
+        # Since connections are no longer rendered with SVG, this is no longer possible.
+        # TODO: add different method of asserting connections
+        pass
+
+    def assert_connection_invalid(self, _source, _sink):
+        # NOOP
+        # TODO: add different method of asserting connections
```
Comment (jmchilton) [inline lib/galaxy_test/selenium/test_workflow_editor.py:1114]: @mvdbeek I know how to replace these other two methods - is the correct way to replace this one checking there is a warning (via https://github.com/galaxyproject/galaxy/pull/15817) upon saving?

**Comment by mvdbeek** [comment]
Panning is a noticeably smoother than on 23.0 now, but moving a single node doesn't look good:

single node movement

**Comment by bgruening** [comment]
This works very well for me and feels even faster than 22.05. 

The only catch I found is that connecting the noodle is a bit hard. You need to exactly drop the noodle to the input green circle. However, I think this is unrelated to this PR. A bit of a snapping mechanism would be nice.

**Comment by mvdbeek** [comment]
It still has the problem in https://github.com/galaxyproject/galaxy/pull/15799#issuecomment-1490074915, that should definitely be fixed before we merge this. 

I would ideally also like to see this approach but with SVG, because I think there are some very useful improvements in here that would also work for SVG.

There is a theoretical performance improvement with rendering to a canvas, but we should be far from reaching this even with large workflows, and being able to introspect SVG has a lot of advantages. See https://blog.logrocket.com/when-to-use-html5s-canvas-ce992b100ee8/, we should be very far in the lower left corner of the render times graph, and if you look at the performance tab all time is spent scripting with the DOM variant.

If you can fix up the single node rendering I'm happy to merge this.

**Comment by ElectronicBlueberry** [comment]
Closed in favor of #16000

**Pull request closed without merging**